File: docs/tests/scala/readers/GeoPackageExamples.scala
package tests.docs.scala.readers

import org.apache.spark.sql.{DataFrame, SparkSession}

/**
  * GeoPackage Reader Examples - Single Source of Truth
  *
  * All Scala code examples shown in docs/docs/readers/geopackage.mdx are defined here.
  * Uses payload-only pattern: object constants for docs display, methods for test validation.
  */
object GeoPackageExamples {

  // Display constants (payload only) - shown in documentation
  val READ_GEOPACKAGE: String =
    """val df = spark.read.format("gpkg_ogr").load("/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg")"""

  /** Example output for READ_GEOPACKAGE (displayed via outputConstant in geopackage.mdx). */
  val READ_GEOPACKAGE_output: String =
    """+--------------------+--------------+---------+
|shape               |shape_srid    |BoroName |
+--------------------+--------------+---------+
|[BINARY]            |4326          |Manhattan|
|...                 |...           |...      |
+--------------------+--------------+---------+"""

  val READ_SPECIFIC_LAYER: String =
    """val boroughs = spark.read.format("gpkg_ogr")
      |  .option("layerName", "boroughs")
      |  .load("/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg")""".stripMargin

  val SQL_GEOPACKAGE: String =
    """SELECT * FROM gpkg_ogr.`/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg`;"""

  // Test methods (validate logic) - used by ScalaTest
  def readGeoPackage(spark: SparkSession, path: String = "/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg"): DataFrame = {
    spark.read.format("gpkg_ogr").load(path)
  }

  def readSpecificLayer(spark: SparkSession, path: String = "/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg", layer: String = "boroughs"): DataFrame = {
    spark.read.format("gpkg_ogr")
      .option("layerName", layer)
      .load(path)
  }
}

File: docs/tests/python/readers/geopackage_examples.py
"""
GeoPackage Reader Examples - Single Source of Truth

All code examples shown in docs/docs/readers/geopackage.mdx are imported from this file.
Uses sample-data Volumes path; output constants for Example output blocks.
"""

# Sample-data path at runtime (path_config)
from path_config import SAMPLE_DATA_BASE
SAMPLE_GEOPACKAGE_PATH = f"{SAMPLE_DATA_BASE}/nyc/geopackage/nyc_complete.gpkg"

# Display constants (payload only)
READ_GEOPACKAGE = """# Read GeoPackage (sample-data Volumes path)
df = spark.read.format("gpkg_ogr").load("/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg")
df.show()"""

READ_GEOPACKAGE_output = """+--------------------+--------------+---------+
|shape               |shape_srid    |BoroName |
+--------------------+--------------+---------+
|[BINARY]            |4326          |Manhattan|
|...                 |...           |...      |
+--------------------+--------------+---------+"""

READ_SPECIFIC_LAYER = """# Read specific layer (sample-data Volumes path)
boroughs = spark.read.format("gpkg_ogr") \\
    .option("layerName", "boroughs") \\
    .load("/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg")
boroughs.show()"""

READ_SPECIFIC_LAYER_output = """+--------------------+--------------+---------+
|shape               |shape_srid    |BoroName |
+--------------------+--------------+---------+
|[BINARY]            |4326          |...      |
|...                 |...           |...      |
+--------------------+--------------+---------+"""

SQL_GEOPACKAGE = """-- Read GeoPackage in SQL (sample-data Volumes path)
SELECT * FROM gpkg_ogr.`/Volumes/main/default/geobrix_samples/geobrix-examples/nyc/geopackage/nyc_complete.gpkg`;"""

SQL_GEOPACKAGE_output = """+--------------------+--------------+---------+
|shape               |shape_srid    |BoroName |
+--------------------+--------------+---------+
|[BINARY]            |4326          |...      |
|...                 |...           |...      |
+--------------------+--------------+---------+"""

# Test functions (validate logic; use SAMPLE_GEOPACKAGE_PATH when path not given)
def read_geopackage(spark, path=None):
    """Verify READ_GEOPACKAGE pattern works."""
    return spark.read.format("gpkg_ogr").load(path or SAMPLE_GEOPACKAGE_PATH)


def read_specific_layer(spark, path=None, layer="boroughs"):
    """Verify READ_SPECIFIC_LAYER pattern works."""
    return spark.read.format("gpkg_ogr").option("layerName", layer).load(path or SAMPLE_GEOPACKAGE_PATH)
